sqldb/invoices.go

```diff
@@ -0,0 +1,1651 @@
+func (i *InvoiceStore) DeleteInvoice(ctx context.Context,
+	invoicesToDelete []invpkg.InvoiceDeleteRef) error {
+
+	// All the InvoiceDeleteRef instances include the add index of the
+	// invoice. The rest was added to ensure that the invoices were deleted
+	// properly in the kv database. When we have fully migrated we can
+	// remove the rest of the fields.
+	for _, ref := range invoicesToDelete {
+		if ref.AddIndex == 0 {
+			return fmt.Errorf("unable to delete invoice using a "+
+				"ref without AddIndex set: %v", ref)
+		}
+	}
+
+	var writeTxOpt InvoiceQueriesTxOptions
+	err := i.db.ExecTx(ctx, &writeTxOpt, func(db InvoiceQueries) error {
+		for _, ref := range invoicesToDelete {
+			params := sqlc.DeleteInvoiceParams{
+				AddIndex: sqlInt64(ref.AddIndex),
+			}
+
+			if ref.SettleIndex != 0 {
+				params.SettledIndex = sqlInt64(ref.SettleIndex)
+			}
+
+			if ref.PayHash != lntypes.ZeroHash {
+				params.Hash = ref.PayHash[:]
+			}
+
+			result, err := db.DeleteInvoice(ctx, params)
+			if err != nil {
+				return fmt.Errorf("unable to delete "+
+					"invoice(%v): %w", ref.AddIndex, err)
+			}
+			rowsAffected, err := result.RowsAffected()
+			if err != nil {
+				return fmt.Errorf("unable to get rows "+
+					"affected: %w", err)
+			}
+			if rowsAffected == 0 {
+				return fmt.Errorf("%w: %v",
+					invpkg.ErrInvoiceNotFound, ref.AddIndex)
+			}
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("unable to delete invoices: %w", err)
+	}
+
+	return nil
+}
+
+// DeleteCanceledInvoices removes all canceled invoices from the database.
+func (i *InvoiceStore) DeleteCanceledInvoices(ctx context.Context) error {
+	var writeTxOpt InvoiceQueriesTxOptions
+	err := i.db.ExecTx(ctx, &writeTxOpt, func(db InvoiceQueries) error {
+		_, err := db.DeleteCanceledInvoices(ctx)
+		if err != nil {
+			return fmt.Errorf("unable to delete canceled "+
+				"invoices: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("unable to delete invoices: %w", err)
+	}
+
+	return nil
+}
+
+// fetchInvoiceData fetches the common invoice data for the given params.
+func fetchInvoiceData(ctx context.Context, db InvoiceQueries,
+	row sqlc.Invoice, setID *[32]byte, fetchAmpHtlcs bool) (*lntypes.Hash,
+	*invpkg.Invoice, error) {
+
+	// Unmarshal the common data.
+	hash, invoice, err := unmarshalInvoice(row)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to unmarshal "+
+			"invoice(id=%d) from db: %w", row.ID, err)
+	}
+
+	// Fetch the invoice features.
+	features, err := getInvoiceFeatures(ctx, db, row.ID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	invoice.Terms.Features = features
+
+	// If this is an AMP invoice, we'll need fetch the AMP state along
+	// with the HTLCs (if requested).
+	if invoice.IsAMP() {
+		invoiceID := int64(invoice.AddIndex)
+		ampState, ampHtlcs, err := fetchAmpState(
+			ctx, db, invoiceID, setID, fetchAmpHtlcs,
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if len(ampHtlcs) > 0 {
```

Comment:
I think this should be done in `fetchAmpState`

Reply:
Good idea, done.